### PR TITLE
Add vyos_interfaces resource module [vyos_interfaces]

### PR DESCRIPTION
**Pull request by NilashishC**
Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

##### SUMMARY
- Add new resource module `vyos_interfaces`.
- Targets model - https://github.com/ansible-network/resource_module_models/pull/4
[vyos_interfaces]
##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
vyos_interfaces.py

### CODE REVIEW

**Review by pabelanger — CHANGES_REQUESTED** [review]
-1 here to ensure we do not merge this code until we discuss the testing strategy. My understanding so far, is our new resource modules require 1.2.x of vyos, but currently we only have 1.1.8 images available.  We can upgrade our vyos image in zuul, but we should confirm if that is something we really want to do.

Currently 1.1.8 is the legacy images, which do not appear to be supported anymore upstream, but assume that users still use them. By switching to 1.2.x, we get the latest images, but due to new license module we cannot run released versions, but we have the ability to use rolling releases (latest builds).

What we need to decided, by switching to 1.2.x, are we also going to continue running 1.1.8 against devel (target for ansible 2.9) or run both images?  By upgrading to 1.2.x, we also need to ensure that all our existing code works with 1.2.x, and deal with any functional changes too.

We would continue to run 1.1.8 on stable-2.8, it would not have the ability to run 1.2.x.

**Comment by NilashishC** [comment]
recheck

File: lib/ansible/modules/network/vyos/vyos_facts.py
Comment (pabelanger) [inline lib/ansible/modules/network/vyos/vyos_facts.py]: We should also indicate 1.1.8 here, or wait on merging until we have 1.2.x online.

**Comment by NilashishC** [comment]
recheck

File: test/integration/targets/vyos_interfaces/tests/cli/overridden.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/tests/cli/overridden.yaml:13]: we shouldn't test against eth0, this is our mgmt interface

File: test/integration/targets/vyos_interfaces/tests/cli/rtt.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/tests/cli/rtt.yaml]: same comment

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:4]: same, lets drop eth0

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:34]: here too

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:83]: another

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:128]: another

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:150]: another

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:186]: here

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (pabelanger) [inline test/integration/targets/vyos_interfaces/vars/main.yaml:208]: here

**Review by pabelanger — COMMENTED** [review]
We should look to see if changed to eth0 are causing issues

File: test/integration/targets/vyos_interfaces/tests/cli/overridden.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/tests/cli/overridden.yaml:13]: Even though eth0 is mentioned, these tests actually do not modify any attributes of eth0. Given the way overridden is designed to work, if we don't specify eth0 in the task, it'll actually end up modifying its attributes. If you look at the commands generated by this task, you will not find any command that concerns eth0.

File: test/integration/targets/vyos_interfaces/tests/cli/rtt.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/tests/cli/rtt.yaml]: Even though eth0 is mentioned, these tests actually do not modify any attributes of eth0. Given the way overridden is designed to work, if we don't specify eth0 in the task, it'll actually end up modifying its attributes. If you look at the commands generated by this task, you will not find any command that concerns eth0.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:4]: This is not a test. This is the fixtures file.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:34]: This is not a test. This is the fixtures file.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:83]: This is not a test. This is the fixtures file.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:128]: This is not a test. This is the fixtures file.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:150]: This is not a test. This is the fixtures file.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:186]: This is not a test. This is the fixtures file.

File: test/integration/targets/vyos_interfaces/vars/main.yaml
Comment (NilashishC) [reply test/integration/targets/vyos_interfaces/vars/main.yaml:208]: This is not a test. This is the fixtures file.

**Comment by NilashishC** [comment]
> We should look to see if changed to eth0 are causing issues

We are not modifying any attributes of eth0 interface in any of the tests. We have specified it in (a) vars/main.yml which contains the fixtures for assertions. (b) In overridden and rtt tests to make sure that eth0 is in now way affected by the aforementioned tests (given the way overridden works).

**Comment by ikhan2010** [comment]
@pabelanger I believe we agreed on the testing strategy.  Do you have any other reservations?

**Comment by ikhan2010** [comment]
@justjais @ganeshrn - Please review the code, so we can merge it.

**Review by pabelanger — COMMENTED** [review]
Nope, testing is working now, on vyos 1.1.8. We still need to bring 1.2.x online, still working to do that.

File: lib/ansible/module_utils/network/vyos/facts/legacy/base.py
Comment (rohitthakur2590) [inline lib/ansible/module_utils/network/vyos/facts/legacy/base.py]: I think it should be VyOS

File: lib/ansible/module_utils/network/vyos/facts/interfaces/interfaces.py
Comment (ganeshrn) [inline lib/ansible/module_utils/network/vyos/facts/interfaces/interfaces.py]: If this is not required it can be removed

File: lib/ansible/module_utils/network/vyos/utils/utils.py
Comment (ganeshrn) [inline lib/ansible/module_utils/network/vyos/utils/utils.py:13]: If this function is used across platforms for resource modules can it be moved to `network/common/utils`

File: lib/ansible/modules/network/vyos/vyos_facts.py
Comment (ganeshrn) [inline lib/ansible/modules/network/vyos/vyos_facts.py]: you can add multiple authors as a list here :-)

File: lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py
Comment (ganeshrn) [inline lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py]: What is the reason to use kwargs here? Since the argument list is defined I don't see a reason to use kwargs. Applicable at other places as well in this file

**Review by ganeshrn — COMMENTED** [review]
Please deprecate the existing vyos_interface module.

File: lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py
Comment (ganeshrn) [inline lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py]: Since replaced is equivalent to deleted of existing attributes in want and after that merged, can this be refactored to reuse the state=deleted logic
Something similar to
```
commands.extend(Interfaces._state_deleted(want, have))
commands.extend(Interfaces._state_merged(want, have))
```

File: lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py
Comment (ganeshrn) [inline lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py]: Similar comment as above.
```
commands.extend(Interfaces._state_replaced(want, have))
commands.extend(Interfaces._state_deleted(<objs in have but not in want>, have))
```

File: lib/ansible/modules/network/vyos/vyos_facts.py
Comment (NilashishC) [reply lib/ansible/modules/network/vyos/vyos_facts.py]: Do you want me to change this to a list?

File: lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py
Comment (rohitthakur2590) [inline lib/ansible/module_utils/network/vyos/config/interfaces/interfaces.py]: aren't we going to handle the scenario where playbook won't have any interface mentioned  and just the
state: deleted which will lead to deleting  the attributes of all the interfaces which were configured by vyos_interfaces
RM.

**Review by pabelanger — APPROVED** [review]
Thanks!

**Review by Qalthos — APPROVED** [review]

**Pull request merged**
